### DISCUSSION TITLE
Location masking change search params

### DISCUSSION THREAD

**Discussion by ekolivero** [TanStack/router#921] [General]
First of all, I'd like to thank you for your efforts in developing this awesome library.

While working on masked routes I've encountered a little problem. When a route is masked we are losing the ability to commit search changes. 

Checking the source code seems like that this behaviour is intentional: https://github.com/TanStack/router/blob/main/packages/react-router/src/router.ts#L886

```tsx
commitLocation = async ({
    startTransition,
    ...next
  }: ParsedLocation & CommitLocationOptions) => {
    if (this.navigateTimeout) clearTimeout(this.navigateTimeout)

    const isSameUrl = this.latestLocation.href === next.href

    // If the next urls are the same and we're not replacing,
    // do nothing
    if (!isSameUrl || !next.replace) {
      let { maskedLocation, ...nextHistory } = next

      if (maskedLocation) {
        nextHistory = {
          ...maskedLocation,
          state: {
            ...maskedLocation.state,
            __tempKey: undefined,
            __tempLocation: {
              ...nextHistory,
              search: nextHistory.searchStr,
              state: {
                ...nextHistory.state,
                __tempKey: undefined!,
                __tempLocation: undefined!,
                key: undefined!,
              },
            },
          },
        }

        if (
          nextHistory.unmaskOnReload ??
          this.options.unmaskOnReload ??
          false
        ) {
          nextHistory.state.__tempKey = this.tempLocationKey
        }
      }

      const apply = () => {
        this.history[next.replace ? 'replace' : 'push'](
          nextHistory.href,
          nextHistory.state,
        )
      }

      if (startTransition ?? true) {
        this.startReactTransition(apply)
      } else {
        apply()
      }
    }

    this.resetNextScroll = next.resetScroll ?? true

    return this.latestLoadPromise
  }
```

However, I've noticed an issue regarding state management in URLs. Specifically, if the route includes a state within the URL, changes are not appropriately reflected when the page is reloaded or when the link is shared

This leads me to a question: does it make sense to give the user the ability to set searchParams in a masked route?

If you guys agree, I can work on a PR. 

Little screenshot to describe desired result.

![image](https://github.com/TanStack/router/assets/20553228/e696c6ba-c1f2-4725-87d4-fb26e76d0ab0)

**Comment by 5ebastianMeier** [comment]
I've come across the same issue.
The route mask only reflects the previous state of the search params, but never the current that is displayed in the DevTools.
If you provide a callback to `search` in the route mask options the only argument you get is `previousState` - so this seems to be by design.
I currently don't understand the reasoning for this so I can only speculate if there's a problem that hasn't been solved yet or if this is the way it's meant to be.
However, having this feature would enable a lot of use cases that are currently not possible.
